[macos] Build process failed without changing anything

**Describe the bug**
I cloned the example folder and ran it in the macOS desktop environment. I closed the application, didn't change anything, and then tried to run it again, but the build failed. If I delete the repository and reclone it, the app runs fine, without changing anything else to my knowledge.

## command

flutter run -d macOS -v

## exception

FileSystemException: FileSystemException: Cannot resolve symbolic links, path = '/Users/me/Desktop/Developing/Flutter/flutter-desktop-embedding/example/macos/Flutter/ephemeral/FlutterMacOS.framework/FlutterMacOS' (OS Error: No such file or directory, errno = 2)
The following build commands failed:
	CompileSwift normal x86_64 /Users/me/Desktop/Developing/Flutter/flutter-desktop-embedding/example/macos/Flutter/GeneratedPluginRegistrant.swift
	CompileSwift normal x86_64 /Users/me/Desktop/Developing/Flutter/flutter-desktop-embedding/example/macos/MainFlutterWindow.swift
	CompileSwift normal x86_64 /Users/me/Desktop/Developing/Flutter/flutter-desktop-embedding/example/macos/AppDelegate.swift
	CompileSwiftSources normal x86_64 com.apple.xcode.tools.swift.compiler
(4 failures)
Finished with error: Build process failed

```
#0      FileSystemEntity._throwIfError (dart:io/file_system_entity.dart:829:7)
#1      FileSystemEntity.resolveSymbolicLinksSync (dart:io/file_system_entity.dart:395:5)
#2      ForwardingFileSystemEntity.resolveSymbolicLinksSync (package:file/src/forwarding/forwarding_file_system_entity.dart:58:49)
#3      writeListIfChanged (package:flutter_tools/src/commands/assemble.dart:152:25)
#4      AssembleCommand.runCommand (package:flutter_tools/src/commands/assemble.dart:140:7)
<asynchronous suspension>
#5      FlutterCommand.verifyThenRunCommand (package:flutter_tools/src/runner/flutter_command.dart:490:18)
<asynchronous suspension>
#6      FlutterCommand.run.<anonymous closure> (package:flutter_tools/src/runner/flutter_command.dart:407:33)
<asynchronous suspension>
#7      AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:154:29)
<asynchronous suspension>
#8      _rootRun (dart:async/zone.dart:1124:13)
#9      _CustomZone.run (dart:async/zone.dart:1021:19)
#10     _runZoned (dart:async/zone.dart:1516:10)
#11     runZoned (dart:async/zone.dart:1463:12)
#12     AppContext.run (package:flutter_tools/src/base/context.dart:153:18)
<asynchronous suspension>
#13     FlutterCommand.run (package:flutter_tools/src/runner/flutter_command.dart:397:20)
#14     CommandRunner.runCommand (package:args/command_runner.dart:197:27)
<asynchronous suspension>
#15     FlutterCommandRunner.runCommand.<anonymous closure> (package:flutter_tools/src/runner/flutter_command_runner.dart:402:21)
<asynchronous suspension>
#16     AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:154:29)
<asynchronous suspension>
#17     _rootRun (dart:async/zone.dart:1124:13)
#18     _CustomZone.run (dart:async/zone.dart:1021:19)
#19     _runZoned (dart:async/zone.dart:1516:10)
#20     runZoned (dart:async/zone.dart:1463:12)
#21     AppContext.run (package:flutter_tools/src/base/context.dart:153:18)
<asynchronous suspension>
#22     FlutterCommandRunner.runCommand (package:flutter_tools/src/runner/flutter_command_runner.dart:356:19)
<asynchronous suspension>
#23     CommandRunner.run.<anonymous closure> (package:args/command_runner.dart:112:25)
#24     new Future.sync (dart:async/future.dart:224:31)
#25     CommandRunner.run (package:args/command_runner.dart:112:14)
#26     FlutterCommandRunner.run (package:flutter_tools/src/runner/flutter_command_runner.dart:242:18)
#27     run.<anonymous closure>.<anonymous closure> (package:flutter_tools/runner.dart:63:22)
<asynchronous suspension>
#28     _rootRun (dart:async/zone.dart:1124:13)
#29     _CustomZone.run (dart:async/zone.dart:1021:19)
#30     _runZoned (dart:async/zone.dart:1516:10)
#31     runZoned (dart:async/zone.dart:1500:12)
#32     run.<anonymous closure> (package:flutter_tools/runner.dart:61:18)
<asynchronous suspension>
#33     AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:154:29)
<asynchronous suspension>
#34     _rootRun (dart:async/zone.dart:1124:13)
#35     _CustomZone.run (dart:async/zone.dart:1021:19)
#36     _runZoned (dart:async/zone.dart:1516:10)
#37     runZoned (dart:async/zone.dart:1463:12)
#38     AppContext.run (package:flutter_tools/src/base/context.dart:153:18)
<asynchronous suspension>
#39     runInContext (package:flutter_tools/src/context_runner.dart:60:24)
<asynchronous suspension>
#40     run (package:flutter_tools/runner.dart:50:10)
#41     main (package:flutter_tools/executable.dart:65:9)
<asynchronous suspension>
#42     main (file:///Users/me/Desktop/Developing/Flutter/flutter/packages/flutter_tools/bin/flutter_tools.dart:8:3)
#43     _startIsolate.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:303:32)
#44     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:172:12)
```

## flutter doctor

```
[✓] Flutter (Channel master, v1.10.2-pre.41, on Mac OS X 10.14.4 18E226, locale en-US)
    • Flutter version 1.10.2-pre.41 at /Users/me/Desktop/Developing/Flutter/flutter
    • Framework revision 8eee93fa64 (3 hours ago), 2019-09-11 09:38:42 -0700
    • Engine revision 7ea9884ab0
    • Dart version 2.5.0 (build 2.5.0-dev.4.0 be66176534)

[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at /Users/me/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-28, build-tools 28.0.3
    • Java binary at: /Users/me/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/183.5692245/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1343-b01)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 10.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 10.2.1, Build version 10E1001
    • CocoaPods version 1.7.5

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 3.4)
    • Android Studio at /Users/me/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/183.5692245/Android Studio.app/Contents
    • Flutter plugin version 33.3.1
    • Dart plugin version 182.5215
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1343-b01)

[✓] IntelliJ IDEA Ultimate Edition (version 2018.3.5)
    • IntelliJ at /Users/me/Applications/JetBrains Toolbox/IntelliJ IDEA Ultimate.app
    • Flutter plugin version 35.3.1
    • Dart plugin version 183.6270

[✓] Connected device (3 available)
    • macOS  • macOS  • darwin-x64     • Mac OS X 10.14.4 18E226
    • Chrome • chrome • web-javascript • Google Chrome 76.0.3809.132
    • Server • web    • web-javascript • Flutter Tools

• No issues found!
```

Relevant parts of the error log:
```
[ +149 ms] Target debug_unpack_macos failed: Exception: Failed to copy framework (exit 1:
[        ] ---
[        ] cp: cannot overwrite directory
```
```
build/Objects-normal/x86_64/MainFlutterWindow.o
[ +353 ms] [FDE path here]/example/macos/MainFlutterWindow.swift:16:8: error: no such
module 'FlutterMacOS'
[        ] import FlutterMacOS
```

--- a/packages/flutter_tools/lib/src/build_system/build_system.dart
+++ b/packages/flutter_tools/lib/src/build_system/build_system.dart
@@ -146,11 +146,11 @@ abstract class Target {
     final File stamp = _findStampFile(environment);
     final List<String> inputPaths = <String>[];
     for (File input in inputs) {
-      inputPaths.add(input.resolveSymbolicLinksSync());
+      inputPaths.add(input.path);
     }
     final List<String> outputPaths = <String>[];
     for (File output in outputs) {
-      outputPaths.add(output.resolveSymbolicLinksSync());
+      outputPaths.add(output.path);
     }
     final Map<String, Object> result = <String, Object>{
       'inputs': inputPaths,

--- a/packages/flutter_tools/lib/src/build_system/build_system.dart
+++ b/packages/flutter_tools/lib/src/build_system/build_system.dart
@@ -502,7 +502,7 @@ class _BuildInstance {
           outputFiles[output.path] = output;
         }
       } else {
-        printStatus('${node.target.name}: Starting');
+        printStatus('${node.target.name}: Starting due to ${node.invalidatedReasons}');
         await node.target.build(environment);
         printStatus('${node.target.name}: Complete');
 
@@ -653,8 +653,13 @@ class Node {
   /// Output file paths from the previous invocation of this build node.
   final Set<String> previousOutputs = <String>{};
 
-  /// Inout file paths from the previous invocation of this build node.
+  /// Input file paths from the previous invocation of this build node.
   final Set<String> previousInputs = <String>{};
+
+  /// One or more reasons why a task was invalidated.
+  ///
+  /// May be empty if the task was skipped.
+  final Set<InvalidedReason> invalidatedReasons = <InvalidedReason>{};
 
   /// Whether this node needs an action performed.
   bool get dirty => _dirty;
@@ -685,6 +690,7 @@ class Node {
       if (fileHashStore.currentHashes.containsKey(absolutePath)) {
         final String currentHash = fileHashStore.currentHashes[absolutePath];
         if (currentHash != previousHash) {
+          invalidatedReasons.add(InvalidedReason.inputChanged);
           _dirty = true;
         }
       } else {
@@ -698,11 +704,13 @@ class Node {
       // output paths changed.
       if (!currentOutputPaths.contains(previousOutput)) {
         _dirty = true;
+        invalidatedReasons.add(InvalidedReason.outputSetChanged);
         // if this isn't a current output file there is no reason to compute the hash.
         continue;
       }
       final File file = fs.file(previousOutput);
       if (!file.existsSync()) {
+        invalidatedReasons.add(InvalidedReason.outputMissing);
         _dirty = true;
         continue;
       }
@@ -711,6 +719,7 @@ class Node {
       if (fileHashStore.currentHashes.containsKey(absolutePath)) {
         final String currentHash = fileHashStore.currentHashes[absolutePath];
         if (currentHash != previousHash) {
+          invalidatedReasons.add(InvalidedReason.outputChanged);
           _dirty = true;
         }
       } else {
@@ -728,9 +737,25 @@ class Node {
     if (sourcesToHash.isNotEmpty) {
       final List<File> dirty = await fileHashStore.hashFiles(sourcesToHash);
       if (dirty.isNotEmpty) {
+        invalidatedReasons.add(InvalidedReason.inputChanged);
         _dirty = true;
       }
     }
     return !_dirty;
   }
+}
+
+/// A description of why a task was rerun.
+enum InvalidedReason {
+  /// An input file has an updated hash.
+  inputChanged,
+
+  /// An output file has an updated hash.
+  outputChanged,
+
+  /// An output file that is expected is missing.
+  outputMissing,
+
+  /// The set of expected output files changed.
+  outputSetChanged,
 }

--- a/packages/flutter_tools/lib/src/build_system/targets/macos.dart
+++ b/packages/flutter_tools/lib/src/build_system/targets/macos.dart
@@ -122,9 +122,12 @@ abstract class UnpackMacOS extends Target {
     if (basePath.endsWith(fs.path.separator)) {
       basePath = basePath.substring(0, basePath.length);
     }
-    final File targetDirectory = environment
+    final Directory targetDirectory = environment
       .outputDir
-      .childFile('FlutterMacOS.framework');
+      .childDirectory('FlutterMacOS.framework');
+    if (targetDirectory.existsSync()) {
+      targetDirectory.deleteSync(recursive: true);
+    }
     final ProcessResult result = await processManager
         .run(<String>['cp', '-R', basePath, targetDirectory.path]);
     if (result.exitCode != 0) {

--- a/packages/flutter_tools/lib/src/build_system/targets/macos.dart
+++ b/packages/flutter_tools/lib/src/build_system/targets/macos.dart
@@ -118,11 +118,13 @@ abstract class UnpackMacOS extends Target {
       throw MissingDefineException(kBuildMode, 'unpack_macos');
     }
     final BuildMode buildMode = getBuildModeForName(environment.defines[kBuildMode]);
-    final String basePath = artifacts.getArtifactPath(Artifact.flutterMacOSFramework, mode: buildMode);
-    final Directory targetDirectory = environment
+    String basePath = artifacts.getArtifactPath(Artifact.flutterMacOSFramework, mode: buildMode);
+    if (basePath.endsWith(fs.path.separator)) {
+      basePath = basePath.substring(0, basePath.length);
+    }
+    final File targetDirectory = environment
       .outputDir
-      .childDirectory('FlutterMacOS.framework');
-
+      .childFile('FlutterMacOS.framework');
     final ProcessResult result = await processManager
         .run(<String>['cp', '-R', basePath, targetDirectory.path]);
     if (result.exitCode != 0) {

--- a/packages/flutter_tools/lib/src/build_system/targets/macos.dart
+++ b/packages/flutter_tools/lib/src/build_system/targets/macos.dart
@@ -118,10 +118,7 @@ abstract class UnpackMacOS extends Target {
       throw MissingDefineException(kBuildMode, 'unpack_macos');
     }
     final BuildMode buildMode = getBuildModeForName(environment.defines[kBuildMode]);
-    String basePath = artifacts.getArtifactPath(Artifact.flutterMacOSFramework, mode: buildMode);
-    if (basePath.endsWith(fs.path.separator)) {
-      basePath = basePath.substring(0, basePath.length);
-    }
+    final String basePath = artifacts.getArtifactPath(Artifact.flutterMacOSFramework, mode: buildMode);
     final Directory targetDirectory = environment
       .outputDir
       .childDirectory('FlutterMacOS.framework');

--- a/packages/flutter_tools/lib/src/build_system/targets/macos.dart
+++ b/packages/flutter_tools/lib/src/build_system/targets/macos.dart
@@ -290,6 +290,7 @@ abstract class MacOSBundleFlutterAssets extends Target {
   @override
   List<Source> get inputs => const <Source>[
     Source.pattern('{PROJECT_DIR}/pubspec.yaml'),
+    Source.pattern('{BUILD_DIR}/App.framework/App'),
     Source.behavior(MacOSAssetBehavior())
   ];
 

--- a/packages/flutter_tools/lib/src/commands/assemble.dart
+++ b/packages/flutter_tools/lib/src/commands/assemble.dart
@@ -149,7 +149,7 @@ void writeListIfChanged(List<File> files, String path) {
   final StringBuffer buffer = StringBuffer();
   // These files are already sorted.
   for (File file in files) {
-    buffer.writeln(file.resolveSymbolicLinksSync());
+    buffer.writeln(file.path);
   }
   final String newContents = buffer.toString();
   if (!file.existsSync()) {

--- a/packages/flutter_tools/test/general.shard/build_system/targets/macos_test.dart
+++ b/packages/flutter_tools/test/general.shard/build_system/targets/macos_test.dart
@@ -85,17 +85,20 @@ void main() {
     for (File input in inputs) {
       input.createSync(recursive: true);
     }
+    // Create output directory so we can test that it is deleted.
+    environment.outputDir.childDirectory(_kOutputPrefix)
+        .createSync(recursive: true);
+
     when(processManager.run(any)).thenAnswer((Invocation invocation) async {
       final List<String> arguments = invocation.positionalArguments.first;
       final String sourcePath = arguments[arguments.length - 2];
       final String targetPath = arguments.last;
-      // Verify there are no trailing `/`.
-      expect(sourcePath, isNot(endsWith(fs.path.separator)));
-      expect(targetPath, isNot(endsWith(fs.path.separator)));
-
       final Directory source = fs.directory(sourcePath);
-      final Directory target = fs.directory(targetPath)
-        ..createSync(recursive: true);
+      final Directory target = fs.directory(targetPath);
+
+      // verify directory was deleted by command.
+      expect(target.existsSync(), false);
+      target.createSync(recursive: true);
 
       for (FileSystemEntity entity in source.listSync(recursive: true)) {
         if (entity is File) {

--- a/packages/flutter_tools/test/general.shard/build_system/targets/macos_test.dart
+++ b/packages/flutter_tools/test/general.shard/build_system/targets/macos_test.dart
@@ -173,6 +173,26 @@ void main() {
     expect(fs.file(precompiledIsolate).existsSync(), false);
   }));
 
+  test('release/profile macOS application updates when App.framework updates', () => testbed.run(() async {
+    fs.file(fs.path.join('bin', 'cache', 'artifacts', 'engine', 'darwin-x64',
+        'vm_isolate_snapshot.bin')).createSync(recursive: true);
+    fs.file(fs.path.join('bin', 'cache', 'artifacts', 'engine', 'darwin-x64',
+        'isolate_snapshot.bin')).createSync(recursive: true);
+    final File inputFramework = fs.file(fs.path.join(environment.buildDir.path, 'App.framework', 'App'))
+        ..createSync(recursive: true)
+        ..writeAsStringSync('ABC');
+
+    await const ProfileMacOSBundleFlutterAssets().build(environment..defines[kBuildMode] = 'profile');
+    final File outputFramework = fs.file(fs.path.join(environment.outputDir.path, 'App.framework', 'App'));
+
+    expect(outputFramework.readAsStringSync(), 'ABC');
+
+    inputFramework.writeAsStringSync('DEF');
+    await const ProfileMacOSBundleFlutterAssets().build(environment..defines[kBuildMode] = 'profile');
+
+    expect(outputFramework.readAsStringSync(), 'DEF');
+  }));
+
   test('release/profile macOS compilation uses correct gen_snapshot', () => testbed.run(() async {
     when(genSnapshot.run(
       snapshotType: anyNamed('snapshotType'),

--- a/packages/flutter_tools/test/general.shard/build_system/targets/macos_test.dart
+++ b/packages/flutter_tools/test/general.shard/build_system/targets/macos_test.dart
@@ -87,9 +87,16 @@ void main() {
     }
     when(processManager.run(any)).thenAnswer((Invocation invocation) async {
       final List<String> arguments = invocation.positionalArguments.first;
-      final Directory source = fs.directory(arguments[arguments.length - 2]);
-      final Directory target = fs.directory(arguments.last)
+      final String sourcePath = arguments[arguments.length - 2];
+      final String targetPath = arguments.last;
+      // Verify there are no trailing `/`.
+      expect(sourcePath, isNot(endsWith(fs.path.separator)));
+      expect(targetPath, isNot(endsWith(fs.path.separator)));
+
+      final Directory source = fs.directory(sourcePath);
+      final Directory target = fs.directory(targetPath)
         ..createSync(recursive: true);
+
       for (FileSystemEntity entity in source.listSync(recursive: true)) {
         if (entity is File) {
           final String relative = fs.path.relative(entity.path, from: source.path);

